pkg/command/plugin_bundle.go

```diff
@@ -249,3 +261,110 @@ func completeGroupsAndVersionForBundleDownload(_ *cobra.Command, _ []string, toC
+	for _, pe := range matchingPluginEntries {
+		for v := range pe.Artifacts {
+			versions = append(versions, v)
+		}
+	}
+	// Sort in ascending order
+	_ = utils.SortVersions(versions)
+
+	// Create the completions in reverse order
+	comps := make([]string, len(versions))
+	for i := range versions {
+		comps[len(versions)-1-i] = fmt.Sprintf("%s:%s", id, versions[i])
+	}
+	return comps
+}
+
+func completionGetPluginsForBundleDownload() ([]*plugininventory.PluginInventoryEntry, error) {
+	// For a download-bundle, we cannot use the DB cache.  This is because
+	// the download-bundle does not use the configured plugin sources.  Instead it
+	// uses the repo specified by the `--image`` flag, or, without the `--image` flag,
+	// it uses the default central repo automatically.
+
+	// We start by downloading the inventory of the required repo.  This is not
+	// very fast, but there isn't much we can do about it.
+
+	var err error
+	tempDBDir, err := os.MkdirTemp("", "")
+	if err != nil {
+		return nil, err
+	}
+	defer os.RemoveAll(tempDBDir)
+
+	// Download the plugin inventory oci image to tempDBDir
+	repoImage := dpbo.pluginDiscoveryOCIImage
+	inventoryFile := filepath.Join(tempDBDir, plugininventory.SQliteDBFileName)
+	if err := imageProcessorForDownloadBundleComp.DownloadImageAndSaveFilesToDir(repoImage, filepath.Dir(inventoryFile)); err != nil {
+		return nil, err
+	}
+
+	// Read the plugin inventory database to read the plugin groups it contains
+	pi := plugininventory.NewSQLiteInventory(inventoryFile, path.Dir(repoImage))
+	pluginEntries, err := pi.GetPlugins(&plugininventory.PluginInventoryFilter{IncludeHidden: true}) // Include the hidden plugin groups during plugin migration
+	if err != nil {
+		return nil, err
+	}
+	return pluginEntries, err
+}
+
+func completePluginIDForBundleDownload(_ *cobra.Command, _ []string, toComplete string) ([]string, cobra.ShellCompDirective) {
+	PluginEntries, err := completionGetPluginsForBundleDownload()
+	if err != nil {
+		return nil, cobra.ShellCompDirectiveNoFileComp
+	}
+	if PluginEntries == nil {
+		comps := cobra.AppendActiveHelp(nil, fmt.Sprintf("There are no plugins in %s", dpbo.pluginDiscoveryOCIImage))
+		return comps, cobra.ShellCompDirectiveNoFileComp
+	}
+
+	if idx := strings.Index(toComplete, ":"); idx != -1 {
+		// The pluginID is already specified before the :
+		// so now we should complete the group version.
+		// Since more recent versions are more likely to be
+		// useful, we tell the shell to preserve the order
+		// using cobra.ShellCompDirectiveKeepOrder
+		pluginID := toComplete[:idx]
+		versionToComplete := toComplete[idx+1:]
+
+		return completeVersionInPluginIDForDownloadBundle(PluginEntries, pluginID, versionToComplete), cobra.ShellCompDirectiveNoFileComp | cobra.ShellCompDirectiveKeepOrder
+	}
+
+	// Complete plugin group names
+	var comps []string
+	for _, p := range PluginEntries {
+		id := plugininventory.PluginToID(p)
+		if strings.HasPrefix(id, toComplete) {
```

Comment:
You might have wondered why we don't check for `HasPrefix()` when dealing with plugin groups.
I should have put a comment to explain.  This is some advanced shell completion stuff 😄 

If we filter on prefix, we are making the assumption that the only completions we should show _must_ start with what the user has typed.  This turns out to be wrong because `zsh` and `fish` are smarter than that: they can match what the user typed on something else than the prefix of the completion, mostly it is matching on substring (fuzzy matching).  Fish can even match on substrings within the _descriptions_ of the completions.

For example you can try doing this with fish:
```
tanzu completion fish | source
# Set the variable so it exists in the config file
tanzu config set env.TANZU_CLI_ADDITIONAL_PLUGIN_DISCOVERY_IMAGES_TEST_ONLY ''

# The below will match the variable name on the _substring_ ADD
tanzu config set ADD<TAB>
```
But for that to work, we need to return all the completions that _could_ match.  So, it is better not to use `HasPrefix()` but instead to return all valid completions.  The shell will then filter appropriately, doing prefix first, and if nothing matches, then trying substring.

P.S. I'm not able to get this concept to work for this PR (modified), I'll have to investigate later

Reply:
Thanks for the explanation. I have updated the PR to not use `HasPrefix()`.